javalanglib/src/main/scala/java/lang/FloatingPointBits.scala

```diff
@@ -197,124 +232,124 @@ private[lang] object FloatingPointBits {
+
+    // Encode
+    val hi = s | (e << hifbits) | rawToInt(f / 0x100000000L.toDouble)
+    val lo = rawToInt(f)
     (hi.toLong << 32) | (lo.toLong & 0xffffffffL)
   }
 
-  @inline private def decodeIEEE754(ebits: Int, fbits: Int,
-      minPositiveValue: scala.Double, s: scala.Boolean, e: Int,
-      f: scala.Double): scala.Double = {
-
-    import Math.pow
+  @inline
+  private def decodeIEEE754(ebits: Int, fbits: Int,
+      powsOf2: js.Array[scala.Double], minPositiveValue: scala.Double,
+      sign: scala.Int, e: Int, f: scala.Double): scala.Double = {
 
     // Some constants
-    val bias = (1 << (ebits - 1)) - 1
     val specialExponent = (1 << ebits) - 1
     val twoPowFbits = (1L << fbits).toDouble
 
-    val absResult = if (e == specialExponent) {
+    if (e == specialExponent) {
       // Special
       if (f == 0.0)
-        scala.Double.PositiveInfinity
+        sign * scala.Double.PositiveInfinity
       else
         scala.Double.NaN
     } else if (e > 0) {
       // Normalized
-      pow(2, e - bias) * (1 + f / twoPowFbits)
+      sign * powsOf2(e) * (1 + f / twoPowFbits)
     } else {
       // Subnormal
-      f * minPositiveValue
+      sign * f * minPositiveValue
     }
+  }
 
-    if (s) -absResult else absResult
+  /** Force rounding of `av` to fit in 32 bits (this is a manual `fround`).
+   *
+   *  `av` must not be negative, i.e., `av < 0.0` must be false (it can be
+   *  `NaN` or `Infinity`).
+   *
+   *  When we use strict-float semantics, this is redundant, because the input
+   *  came from a `Float` and is therefore guaranteed to be rounded already.
+   *  However, here we don't know whether we use strict floats semantics or
+   *  not, so we must always do it. This is not a big deal because, if this
+   *  code is called, then any operation on `Float`s is calling the same code
+   *  from the `CoreJSLib`, so doing one more such operation for
+   *  `floatToIntBits` is negligible.
+   *
+   *  TODO Remove this when we get rid of non-strict float semantics altogether.
+   */
+  @inline
+  private def forceFround(av: scala.Double): scala.Double = {
+    // See the `fround` polyfill in CoreJSLib
+    val overflowThreshold = 3.4028235677973366e38
+    val normalThreshold = 1.1754943508222875e-38
+    if (av >= overflowThreshold) {
+      scala.Double.PositiveInfinity
+    } else if (av >= normalThreshold) {
+      val p = av * 536870913.0 // pow(2, 29) + 1
+      p + (av - p)
+    } else {
+      val roundingFactor = scala.Double.MinPositiveValue / scala.Float.MinPositiveValue.toDouble
+      (av * roundingFactor) / roundingFactor
+    }
   }
 
-  @inline private def encodeIEEE754(ebits: Int, fbits: Int,
-      minNormal: scala.Double, minPositiveValue: scala.Double,
-      v: scala.Double): EncodeIEEE754Result = {
+  @inline
+  private def encodeIEEE754Exponent(ebits: Int,
+      powsOf2: js.Array[scala.Double], av: scala.Double): Int = {
```

Comment:
What is the reason this is `@inline`? Before, it was because of the return type, but now?

Also, this comment somewhat applies as well to `decodeIEEE754` but there it would have applied even before this commit.

Reply:
For `decodeIEEE754` it's important for its `ebits` and `fbits` parameter to be constant-folded, which in turns makes the "`// Some constants`" actually constant. If we don't, the constants need to be recomputed every time, which even includes `Long` operations (!).

For this method, there is no such strong argument. I can try and remove it and redo the benchmarks, if you want.

Reply:
> For decodeIEEE754 it's important for its ebits and fbits parameter to be constant-folded, which in turns makes the "// Some constants" actually constant. If we don't, the constants need to be recomputed every time, which even includes Long operations (!).

Ah, that makes sense.

> For this method, there is no such strong argument. I can try and remove it and redo the benchmarks, if you want.

If it is not too much effort, maybe? It just seems strange to inline a larger method like this. OTOH, not sure it is really relevant in terms of code size.

I'll leave it up to you.[

Reply:
I redid the benchmarks, and basically the results are indistinguishable between with `@inline` and without. So I removed it.